### ISSUE TITLE
[question] client connection to nowhere results into Ok/Success, how to check real state?

### ISSUE THREAD

**Issue by denveloper** [pingcap/grpc-rs#228]
Hi,
I have the following code:

```
let cb = ChannelBuilder::new(Arc::new(EnvBuilder::new().build()))
let ch = cb.connect("localhost:33333");
let client = PeerClient::new(ch);

match client.exchange() {
    Ok((sender, receiver)) => {
        // Getting here               
    }
    Err(e) => {
        // Expect to be here because no sever on port 33333
    }
}

```

This is a client for bi-directional streaming (grpc.proto: 
```
rpc Exchange(stream Data) returns (stream Data);
```
). 

There is nothing on port 33333 but 'client.exchange' still results in Ok. Is there a way to check that connection was successful before sending anything to the channel?

In my application, I have separate modules: #1 client creation module and #2 stream processing module. I would like to not propagate a bogus event that there is a connection where there is none.

If I do receiver.into_future().wait() I am getting an error immediately but if the connection is correct I also have "data" that is no use in the module that handles client creation.

So, in other words, is there way to check that "client.exchange()" was successful without pumping data to/from?

Thank you!

**Comment by BusyJay** [comment]
gRPC provides [APIs](https://github.com/grpc/grpc/blob/master/doc/connectivity-semantics-and-api.md) to detect the state of a channel. But they are not included in this crate yet. We are also open for pull requests if you want to add it by yourself.
